File: client/src/main/scala/stasis/client/ops/commands/CommandProcessor.scala
package stasis.client.ops.commands

import scala.concurrent.Future

import org.apache.pekko.Done

import stasis.core.commands.proto.Command

trait CommandProcessor {
  def all(): Future[Seq[Command]]
  def latest(): Future[Seq[Command]]
  def stop(): Future[Done]

  def lastProcessedCommand: Future[Option[Long]] = handlers.retrieveLastProcessedCommand()

  protected def handlers: CommandProcessor.Handlers
}

object CommandProcessor {
  trait Handlers {
    def persistLastProcessedCommand(sequenceId: Long): Future[Done]
    def retrieveLastProcessedCommand(): Future[Option[Long]]
    def executeCommands(commands: Seq[Command]): Future[Long]
  }
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/ops/commands/CommandProcessor.kt
package stasis.client_android.lib.ops.commands

import stasis.client_android.lib.utils.Try
import stasis.core.commands.proto.Command

interface CommandProcessor {
    suspend fun all(): Try<List<Command>>
    suspend fun latest(): Try<List<Command>>
    suspend fun stop()

    interface Handlers {
        suspend fun persistLastProcessedCommand(sequenceId: Long)
        suspend fun retrieveLastProcessedCommand(): Long
        suspend fun executeCommands(commands: List<Command>): Long?
    }
}
